src/main/java/net/neoforged/neoforge/event/entity/EntityInvulnerabilityCheckEvent.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event.entity;
+
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Fired when {@link Entity#isInvulnerableTo(DamageSource)} is invoked and determines if
+ * downstream hurt logic should apply. This event is fired on both sides in
+ * {@link Entity#isInvulnerableTo(DamageSource)}
+ * <br>
+ * <b>Note: This event may be unable to change the invulnerable status of some entities
+ * that override isInvulnerableTo against certain damage sources</b>
```

Comment:
I noticed that Ghast fireballs that users hits reflected back will cause this event to not fire so you can't make Ghasts immune to reflected fireballs. Is there any way we can make this event a bit more useful? Or is the fear that patching every entity that overrides isInvulnerableTo to fire the event first, something that is too much to do?

Wondering if it could be an asm check that wraps the entirety of all isInverulerableTo methods into the onEntityInvulnerabilityCheck method. But then the event could fire multiple times for a single entity that calls super...

Annoying.

Reply:
So we aren't really concerned about multiple firing of the invulnerability check.  That came up in discussion.  If there is a clean way to capture those invulnerability checks with overrides using a fancy ASM, i welcome the input.  

Ultimately you are right about why we didn't inject into these entities.  There are 3 entities with special cases that result in the invulnerability check not firing: Breezes, Ghasts, and the Warden.  Anyone who needs to circumvent the vanilla behavior for these cases can address that themselves, but for the sake of keeping this feature clean, they were left out.

Reply:
That may be an avenue to explore in the future, but is currently blocked by https://github.com/neoforged/NeoForge/issues/1168